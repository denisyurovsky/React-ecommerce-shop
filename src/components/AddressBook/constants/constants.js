export const LISTBOX_PADDING = 8;
export const DOUBLE_LISTBOX_PADDING = LISTBOX_PADDING * 2;
export const LIMIT_NUMBER_ITEMS = 8;
export const ITEM_SIZE_LG = 48;
export const ITEM_SIZE_MD = 36;
