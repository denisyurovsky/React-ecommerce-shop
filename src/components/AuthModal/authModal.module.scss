@import "../../assets/styles/colors.scss";

.container {
  display: flex;
  justify-content: center;
  align-items: center;
  width: 100vw;
  height: 100vh;
}

.modal {
  position: relative;
  width: 40%;
  max-width: 600px;
  min-width: 300px;
  box-shadow: 0 0 20px $grey700;
  background-color: $white;
  border-radius: 4px;
}

.cross {
  position: absolute !important;
  top: 5px;
  right: 5px;
}

.forms {
  padding: 40px;
  padding-top: 16px;
}

@media (max-width: 575px) {
  .forms {
    padding: 16px;
  }
}
