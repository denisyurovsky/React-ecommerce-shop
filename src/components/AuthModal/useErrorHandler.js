import { useEffect, useState } from 'react';

const useErrorHandler = (isFailed) => {
  const [isFailedLocal, setIsFailedLocal] = useState(false);
  const isError = isFailed && isFailedLocal;
  const resetError = () => setIsFailedLocal(false);

  useEffect(() => {
    setIsFailedLocal(isFailed);
  }, [isFailed]);

  return { isError, resetError };
};

export default useErrorHandler;
