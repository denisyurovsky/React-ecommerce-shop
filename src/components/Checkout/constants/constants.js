export const PANEL = {
  DELIVERY_ADDRESS: 'Delivery Address',
  PERSONAL_INFORMATION: 'Personal Information',
  PERSONAL_ADDRESS: 'Personal Address',
  PAYMENT_METHOD: 'Payment method',
  ORDER_CONFIRMATION: 'Order confirmation',
};
