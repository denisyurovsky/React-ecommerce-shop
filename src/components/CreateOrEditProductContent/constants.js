export const PRICE_ERROR = 'Only numbers';
export const DESCRIPTION_ERROR = 'Only latin characters';
export const LENGTH_ERROR = 'Should contain at least one symbol';
export const DISCOUNT_ERROR = 'Discount should be less than price';
export const SELECTION_ERROR = 'Choose category';
