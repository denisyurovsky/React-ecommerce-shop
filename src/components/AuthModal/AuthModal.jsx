import CloseIcon from '@mui/icons-material/Close';
import { Tabs, Tab, Box, Modal, IconButton, Button } from '@mui/material';
import PropTypes from 'prop-types';
import React, { useState } from 'react';

import { authStatus } from '../../helpers/constants/authConstants';

import Registration from './Registration/Registration';
import SignIn from './SignIn/SignIn';

import styles from './authModal.module.scss';

const { FULFILLED, IDLE, LOADING, FAILED, LOCKED } = authStatus;

const AuthModal = ({
  registerUser,
  registerStatus,
  loginUser,
  loginStatus,
  errorMessage,
  resetError,
}) => {
  const [value, setValue] = useState('old');
  const [isOpen, setIsOpen] = useState(false);

  const handleChange = (event, newValue) => setValue(newValue);
  const handleClose = () => setIsOpen(false);
  const handleOpen = () => setIsOpen(true);

  const isLogged = loginStatus === FULFILLED || registerStatus === FULFILLED;

  const open =
    isOpen && loginStatus !== FULFILLED && registerStatus !== FULFILLED;

  return (
    <>
      {isLogged ? (
        <Button
          data-testid="btn-profile"
          className={styles.button}
          variant="outlined"
        >
          Profile
        </Button>
      ) : (
        <Button
          data-testid="btn-login"
          onClick={handleOpen}
          className={styles.button}
          variant="outlined"
        >
          Sign in
        </Button>
      )}
      <Modal open={open} onClose={handleClose} className={styles.container}>
        <Box className={styles.modal}>
          <Box>
            <Tabs
              centered
              value={value}
              onChange={handleChange}
              textColor="primary"
              indicatorColor="primary"
            >
              <Tab
                label="Sign in"
                value="old"
                sx={{ width: 0.25, minWidth: 100 }}
                onClick={resetError}
              />
              <Tab
                label="Sign up"
                value="new"
                sx={{ width: 0.25, minWidth: 100 }}
                onClick={resetError}
              />
            </Tabs>
            <IconButton
              className={styles.cross}
              color="primary"
              onClick={handleClose}
              size="small"
            >
              <CloseIcon />
            </IconButton>
          </Box>
          <Box className={styles.forms}>
            {value === 'new' ? (
              <Registration
                sendForm={registerUser}
                status={registerStatus}
                errorMessage={errorMessage}
              />
            ) : (
              <SignIn
                sendForm={loginUser}
                status={loginStatus}
                errorMessage={errorMessage}
              />
            )}
          </Box>
        </Box>
      </Modal>
    </>
  );
};

AuthModal.defaultProps = {
  errorMessage: '',
  registerStatus: IDLE,
  loginStatus: IDLE,
};

AuthModal.propTypes = {
  resetError: PropTypes.func.isRequired,
  registerUser: PropTypes.func.isRequired,
  loginUser: PropTypes.func.isRequired,
  registerStatus: PropTypes.oneOf([IDLE, LOADING, FULFILLED, FAILED]),
  loginStatus: PropTypes.oneOf([IDLE, LOADING, FULFILLED, FAILED, LOCKED]),
  errorMessage: PropTypes.string,
};

export default AuthModal;
