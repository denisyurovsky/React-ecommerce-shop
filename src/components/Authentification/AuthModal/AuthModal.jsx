import { Tabs, Tab, Box, Button } from '@mui/material';
import PropTypes from 'prop-types';
import React, { useState } from 'react';

import { authStatus } from '../../../constants/authConstants';
import Modal from '../../ui-kit/Modal/Modal';

import Registration from './Registration/Registration';
import SignIn from './SignIn/SignIn';

import styles from './authModal.module.scss';

const { FULFILLED, IDLE, PENDING, REJECTED, LOCKED } = authStatus;

const AuthModal = ({
  registerUser,
  registerStatus,
  loginUser,
  loginStatus,
  loginErrorMessage,
  registerErrorMessage,
  resetError,
}) => {
  const [value, setValue] = useState('old');
  const [isOpen, setIsOpen] = useState(false);

  const handleChange = (event, newValue) => {
    resetError();
    setValue(newValue);
  };
  const handleClose = () => {
    resetError();
    setIsOpen(false);
  };
  const handleOpen = () => setIsOpen(true);

  const open =
    isOpen && loginStatus !== FULFILLED && registerStatus !== FULFILLED;

  return (
    <>
      <Button
        data-testid="btn-login"
        onClick={handleOpen}
        variant="outlined"
        color="secondary"
      >
        Sign in
      </Button>
      <Modal isOpen={open} onClose={handleClose}>
        <Tabs
          centered
          value={value}
          onChange={handleChange}
          textColor="primary"
          indicatorColor="primary"
          className={styles.navigation}
        >
          <Tab
            label="Sign in"
            value="old"
            sx={{ width: 0.25, minWidth: 100 }}
          />
          <Tab
            label="Sign up"
            value="new"
            sx={{ width: 0.25, minWidth: 100 }}
          />
        </Tabs>
        <Box className={styles.content}>
          {value === 'new' ? (
            <Registration
              handleClose={handleClose}
              sendForm={registerUser}
              status={registerStatus}
              errorMessage={registerErrorMessage}
            />
          ) : (
            <SignIn
              sendForm={loginUser}
              status={loginStatus}
              errorMessage={loginErrorMessage}
            />
          )}
        </Box>
      </Modal>
    </>
  );
};

AuthModal.defaultProps = {
  registerErrorMessage: '',
  loginErrorMessage: '',
  registerStatus: IDLE,
  loginStatus: IDLE,
};

AuthModal.propTypes = {
  resetError: PropTypes.func.isRequired,
  registerUser: PropTypes.func.isRequired,
  loginUser: PropTypes.func.isRequired,
  registerStatus: PropTypes.oneOf([IDLE, PENDING, FULFILLED, REJECTED]),
  loginStatus: PropTypes.oneOf([IDLE, PENDING, FULFILLED, REJECTED, LOCKED]),
  loginErrorMessage: PropTypes.string,
  registerErrorMessage: PropTypes.string,
};

export default AuthModal;
