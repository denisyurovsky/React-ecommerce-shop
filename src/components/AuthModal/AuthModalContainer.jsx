import React from 'react';
import { useDispatch, useSelector } from 'react-redux';

import { authStatus } from '../../helpers/constants/authConstants';
import {
  getLoginState,
  getRegisterState,
  loginUser,
  registerUser,
  removeLockAfterTimeout,
  resetBlockedState,
} from '../../store/user/userSlice';

import AuthModal from './AuthModal';

const AuthModalContainer = () => {
  const { LOCKED } = authStatus;
  const dispatch = useDispatch();
  const loginStatus = useSelector(getLoginState);
  const registerStatus = useSelector(getRegisterState);
  const message = useSelector((state) => state.user.error);

  const sendLoginForm = ({ password, email }) => {
    dispatch(loginUser({ password, email }));
  };

  const sendRegisterForm = ({ email, password, ...info }) => {
    dispatch(registerUser({ email, password, ...info }));
  };

  if (loginStatus === LOCKED) {
    dispatch(removeLockAfterTimeout());
  }

  const resetError = () => dispatch(resetBlockedState());

  return (
    <AuthModal
      registerUser={sendRegisterForm}
      registerStatus={registerStatus}
      loginUser={sendLoginForm}
      loginStatus={loginStatus}
      errorMessage={message}
      resetError={resetError}
    />
  );
};

export default AuthModalContainer;
