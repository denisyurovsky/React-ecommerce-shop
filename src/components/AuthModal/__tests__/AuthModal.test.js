import { render, fireEvent, screen } from '@testing-library/react';
import React from 'react';

import { authStatus } from '../../../helpers/constants/authConstants';
import AuthModal from '../AuthModal';

const { FULFILLED } = authStatus;

const openModal = () => {
  const modalBtn = screen.getByTestId('btn-login');

  fireEvent.click(modalBtn);
};

describe('AuthModal snapshot', () => {
  it('should render a valid snapshot', () => {
    const { asFragment } = render(
      <AuthModal
        registerUser={() => {}}
        loginUser={() => {}}
        resetError={() => {}}
      />
    );

    expect(asFragment()).toMatchSnapshot();
  });
});

describe('Navigation functionality tests', () => {
  beforeEach(() => {
    render(
      <AuthModal
        registerUser={() => {}}
        loginUser={() => {}}
        resetError={() => {}}
      />
    );
    openModal();
  });

  it('should open a SignUp forms', () => {
    const navLoginBtn = screen.getByRole('tab', { selected: true });

    expect(navLoginBtn).toHaveAttribute('aria-selected', 'true');
  });

  it('should switch between SignUp and SigIn', () => {
    const navRegisterBtn = screen.getByRole('tab', { selected: false });
    const navLoginBtn = screen.getByRole('tab', { selected: true });

    expect(navLoginBtn).toHaveAttribute('aria-selected', 'true');
    expect(navRegisterBtn).toHaveAttribute('aria-selected', 'false');

    fireEvent.click(navRegisterBtn);
    expect(navLoginBtn).toHaveAttribute('aria-selected', 'false');
    expect(navRegisterBtn).toHaveAttribute('aria-selected', 'true');
  });
});

describe('Closing functionality', () => {
  it('should close after closeIcon is pressed', () => {
    const { getByTestId } = render(
      <AuthModal
        loginUser={() => {}}
        registerUser={() => {}}
        resetError={() => {}}
      />
    );

    openModal();
    const cross = getByTestId('CloseIcon');

    fireEvent.click(cross);

    const profileBtn = getByTestId('btn-login');

    expect(profileBtn).toBeInTheDocument();
  });

  it('should close after successful login', () => {
    const { getByTestId } = render(
      <AuthModal
        loginUser={() => {}}
        registerUser={() => {}}
        loginStatus={FULFILLED}
        resetError={() => {}}
      />
    );
    const profileBtn = getByTestId('btn-profile');

    expect(profileBtn).toBeInTheDocument();
  });

  it('should close after successful registration', () => {
    const { getByTestId } = render(
      <AuthModal
        loginUser={() => {}}
        registerUser={() => {}}
        registerStatus={FULFILLED}
        resetError={() => {}}
      />
    );
    const profileBtn = getByTestId('btn-profile');

    expect(profileBtn).toBeInTheDocument();
  });
});
