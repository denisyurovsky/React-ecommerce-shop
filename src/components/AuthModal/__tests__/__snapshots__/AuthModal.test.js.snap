// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AuthModal snapshot should render a valid snapshot 1`] = `
<DocumentFragment>
  <button
    class="MuiButton-root MuiButton-outlined MuiButton-outlinedPrimary MuiButton-sizeMedium MuiButton-outlinedSizeMedium MuiButtonBase-root button css-1rwt2y5-MuiButtonBase-root-MuiButton-root"
    data-testid="btn-login"
    tabindex="0"
    type="button"
  >
    Sign in
    <span
      class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
    />
  </button>
</DocumentFragment>
`;
